### DISCUSSION TITLE
file upload example

### DISCUSSION THREAD

**Discussion by mc0110** [miguelgrinberg/microdot#114] [Q&A]
Hi Miguel,

it's Magnus once again. Thanks for the great framework.

Your upload example works in the threat variant in my code too, but not in the asyncio variant. Do you have an idea why? The error is: TypeError: object with buffer protocol required.

I have another question. Why can't I use global variables in the code that I need to query / set between requests.

**Comment by miguelgrinberg** [comment]
I cannot comment on code that I haven't seen.

But in any case, I've added an asyncio version of the file upload example: https://github.com/miguelgrinberg/microdot/blob/main/examples/uploads/uploads_async.py.

**Comment by mc0110** [comment]
That is really true! 

I'll try it with your example. If I get stuck, I'll post my code example.

**Comment by mc0110** [comment]
Can you give me a hint on the global variables question - it's a mystery to me

**Comment by mc0110** [comment]
It works now!!
Is it true that the magic line 
`Request.max_content_length = 1024 * 1024 # 1MB (change as needed)`
is? Could you please explain this?

**Comment by miguelgrinberg** [comment]
I can't answer your global variables questions, because I don't understand what you are asking me. Same as before, without code it is difficult. Microdot has nothing to do with global variables, you should be able to use them when it makes sense.

The `max_content_length` is the maximum size you want to allow for uploads. You don't have to ask me about this, by the way, this is all documented: https://microdot.readthedocs.io/en/latest/api.html#microdot_asyncio.Request.max_body_length

**Comment by mc0110** [comment]
I agree to 100 percent - thank you anyway for the quick answer. 

It is sometimes not so easy to use the documentation correctly in a new area. But I can assure you that I have read the document several times. Unfortunately, I am not getting anywhere with two issues:
1) After uploading I would like to jump back to the previous path, but unfortunately I end up on the start page, no matter what I specify as response. I tried "redirect, response, request_request".
2) I unfortunately lose the cookies on "POST" returns. What possibility do you recommend to pass on data even with POST?

        from microdot_asyncio import Microdot, send_file, Request, Response, redirect
        from gen_html import handleRoot, handleHeader, handleFooter, handleFiles 
        import os
        
        app = Microdot()
        Request.max_content_length = 1024 * 1024  # 1MB (change as needed)
        Response.default_content_type = 'text/html'

        @app.post('/upload/<path:dir>')
        async def upload(request, dir):
            print("dir: "+dir)
            if dir == "__/":
                dir = "/"
            print(request.headers)
            # obtain the filename and size from request headers
            filename = request.headers['Content-Disposition'].split('filename=')[1].strip('"')
            size = int(request.headers['Content-Length'])
            # sanitize the filename
            filename = filename.replace('/', '_')
            # write the file to the files directory in 1K chunks
            with open(dir + filename, 'wb') as f:
                while size > 0:
                    chunk = await request.stream.read(min(size, 1024))
                    f.write(chunk)
                    size -= len(chunk)
            f.close()        
            @request.after_request
            def setresponse(request, response):
                response = Response(handleFiles(dir))
                print('Successfully saved file: ' + filename)
                return response
            return ''

I'm not sure if you need more of the code? This was my last try, breaks with a "TypeError: 'Response' object isn't iterable", the other tries run without error - but landing in the root-tree without path.

**Comment by miguelgrinberg** [comment]
> After uploading I would like to jump back to the previous path, but unfortunately I end up on the start page, no matter what I specify as response. I tried "redirect, response, request_request".

What does "previous path" mean here? How would the server know what the previous path is? All you need to do after you save the file is to return a redirect to the place you want to go. The after_request thing that you are doing does not really make sense in this case.

> I unfortunately lose the cookies on "POST" returns. What possibility do you recommend to pass on data even with POST?

What cookies? You don't have any cookies in the code you copied in your question. Cookies cannot be "lost" because they are kept by the client. Nothing you can do on the server other than explicitly telling the client to delete can make a cookie disappear.
